### ISSUE TITLE
contains for hashset finds Object which is changed

### ISSUE THREAD

**Issue by pDiller** [joel-costigliola/assertj-core#1104]
#### Summary
The contains method of HashSet doesn´t find an object when a value changed which is used in the hashcode method of the object. Normally 'assertThat(anyHashSet).contains(changedObject)' should not find it. I wrote a Test and added it here. The Bug appeard in Version 3.8.0. 

#### Example

```java
    @Test
    public void hashSetCanFailToFindTheSameElementPreviouslyAdded() {
        Set<Date> hashSet = new HashSet<>();
        Date date = new Date();
        hashSet.add(date);
        date.setTime(123L);
        Assertions.assertThat(hashSet).contains(date);
        Assertions.assertThat(hashSet.contains(date)).isFalse();
    }
```

#### Java 8 specific ?

* NO  : create Pull Request from the `2.x` branch

**Comment by joel-costigliola** [comment]
That's an interesting one since the [javadoc for HashSet `contains`](https://docs.oracle.com/javase/8/docs/api/java/util/HashSet.html#contains-java.lang.Object-) states:
```
Returns true if this set contains the specified element. More formally, returns true if and only if this set contains an element e such that (o==null ? e==null : o.equals(e)).
```
It turns out that the set in your example contains `date` as per this successful assertion: 
```java
assertThat(hashSet.iterator().next().equals(date)).isTrue();
```
I quickly searched online but could not find whether `HashSet.contains` implementation is considered a bug or just the javadoc which is wrong.

AssertJ actually looks into the collection to state whether an element is contained or not, this is why the assertions does not fail. 

Is that the desired behavior  ? On one hand user would expect AssertJ `contains` assertion to follow `contains` but on the other hand the set actually contains the given element. No clear winner.

I believe the real problem is that one should just not change elements added to a Set, this is a bad practice.

I'm not yet convinced to change that, I'd like to hear opinions.
@PascalSchumacher WDYT ?

**Comment by pDiller** [comment]
Thank you for your fast reply. I think there is an inconsistency between JavaDoc of `HashSet#contains` and the method implementation. `HashSet#contains` is just delegating the call to its internal HashMap. On this `containsKey()` is called, in there the objects hash is build and with this `getNode` is called. So I think the JavaDoc there is misleading. 

But I also think, there are many usecases where you implement your own equals and hashcode methods and maybe you want to change fields, which are used in this methods. 

For sure we should tell the oracle developers that the JavaDoc is not correct, because there is no equals operations directly on the `HashSet#contains`-method. I will open a bugreport on the JDK. If you don´t want to change your api behavior, maybe you can leave a comment on your code? :-)

**Comment by filiphr** [comment]
I am not sure if this is an actual bug in JDK. See [this SO](https://stackoverflow.com/questions/8805807/workaround-for-hash-for-a-hashset-when-the-internal-object-changes) that actually searches for a workaround for changing hash of an object.

The JavaDoc of `HashSet#contains` is correct, as it uses `HashMap#contains` and it uses equals. However, in `HasMap` and `HashSet` implementations `equals` is used only when there is a hash collision. In this case there is no object that matches the new hash.

The reason why this inconsistence exists in AssertJ is that everything is reduced down to an `Iterable`, which is then iterated and an `equals` check is done to check that an object is contained or not. Maybe we need to check if an `Iterable` is actually a `Collection` and call the collection specific methods for that. This way we can be sure that the correct JDK expectations are used.

And finally I also agree with @joel-costigliola one should not change hash of elements added to a `HashSet` and `HashMap`, as it can lead to a lot of inconsistent behaviour.

**Comment by PascalSchumacher** [comment]
I think there is no clean way to fix this.

If we would change AssertJ to use `Collection#equals` we fix this case but break other cases, as AssertJ use the registered `ComparisonStrategy` (`StandardComparisonStrategy` by default, which compares arrays by their contents) to compare the elements contained in the `Iterable`.

I agree that changing the hashcode of elements added to a `HashSet` seems like a bad practice.

**Comment by joel-costigliola** [comment]
Still haven't made my mind, will give it more time.

**Comment by pDiller** [comment]
Maybe I can show you my problem with this issue, on this example:

We have a piece of code:
```
public Set<Date> getADateInASet(Date date) {
        Set<Date> hashSet = new HashSet<>();
        hashSet.add(date);
        return hashSet;
    }
```

If someone changes it unexpectedly to this:

```
public Set<Date> getADateInASet(Date date) {
        Set<Date> hashSet = new HashSet<>();
        hashSet.add(date);
        date.setTime(123L); // unexpected change (yes, bad practice, but can happen)
        return hashSet;
    } 
```

my test:

```
@Test
    public void testDoesNotNoticeThatProductionCodeHasChanged() throws Exception {
        HashSetWrapper a = new HashSetWrapper();
        Date date = new Date();
        Set<Date> hashSet = a.getADateInASet(date);

        Assertions.assertThat(hashSet).contains(date); //For second code example, this should fail
    }
```

should fail, because the code will definately fail in production. This issue is not detected before someone executes the code. I don´t appreciate this.
Shouldn´t this be the reason to write tests? Something should fail in tests aswell, if it fails in production.

**Comment by joel-costigliola** [comment]
Yeah, that's a point, we might go for the change and follow the Java HashSet behavior which is the main argument for me.
PR welcome, with all the people liking this issue it should be easy to find contributors ;-), let me know if one of you want to contribute I need to think at the best approach first (likely defining HashSet or Set specific assertions).

**Comment by nach-o-man** [comment]
I can grab it.

**Comment by joel-costigliola** [comment]
@nach-o-man any update on this ? (it's alright if you are too busy to work on it).

**Comment by nach-o-man** [comment]
@joel-costigliola haven't cheсked it really cause didnt have time. As of right now I will start looking.

**Comment by nach-o-man** [comment]
@joel-costigliola couldnt really find good solution untill now. Will be leaving for vacation untill 24.05 so if anyone feels like grabbing it...

**Comment by joel-costigliola** [comment]
@nach-o-man no problem, thanks for the time you have spent!

**Comment by ardmn** [comment]
Guys, I think it is no problem of assertj :) 

I think that problem in the using of HashSet and HashMap and so on . Please see my explanation in code sample bellow:

```java
import org.junit.Test;

import java.util.Date;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import static org.junit.Assert.*;

public class SetTest {

    @Test
    public void hashSetTest() {
        final Set<Date> set = new HashSet<>();
        final Date date = new Date();

        // add date to set
        set.add(date);

        // set should contains date
        assertTrue(set.contains(date));

        //then change date
        date.setTime(123L);

        //I think that set should still contain date. Really Isn't it so ?
        //but set.contains returns false
        assertFalse(set.contains(date));

        // How ? Let's try to get all set's elements and check each.
        assertTrue(isContains(set, date));
        //We were able find it :)

        /*
         Resume.
         HashSet(HashMap)'s  method 'contains' can't find the date object but set
         has reference to it and if we will iterated on set then we can find it.
         */

        // then try do next: add to set our date object again.
        set.add(date);

        assertEquals(2, set.size());

        //really ? How set can contain 2 equals objects ? ;)
       // Objects  got into different backed of HashMap that using for implementation of HashSet 

        /*
         Resume.
         In relation to work specificity of HashSet,HashMap and other classes and in relation to contract of
         equals-hashCode method user can't change object's fields which effects results of 'equals' and 'hashCode'
         methods.

         More detail about work of HashMap u can find for example in http://www.javainterviewpoint.com/hashmap-works-internally-java/

         And I commend watch this video: http://jqno.nl/equalsverifier/ and use EqualsVerifier :)
         */

    }

    private boolean isContains(final Set<Date> set, final Date searchingDate) {
        for (Date date : set) {
            if (Objects.equals(date, searchingDate)) {
                return true;
            }
        }
        return false;
    }

}

```

**Comment by pDiller** [comment]
@ardmn Please look at the iterator-method of HashSet. It delegates the call to the internal HashMap, so you get all keys from the map. Of course you can find the changed object when you iterate over the keys, because its in there. But the contains-method of HashSet delegates to HashMap#containsKey, which builds a new hash for the given object. This hash is not equals the hash you put in there before. This is the reason HashSet can contains 2 elements. They aren´t equal. ;-) I´m sorry but i think your argument is not valid.

**Comment by pDiller** [comment]
@ardmn Sorry for the late answer. I think the main problem is, that iterating the values is not possible for hashset. Hashsets contract is that all values are hidden behind the hashvalue of the value. If you make an equals comparison by iterating over the values you can find it without a problem, the value and the hash of the value is in the internal hashmap. BUT, if you change the value, the hash of the value is changed, so you find the value by iterating but nether by hashcomparison, which is the official contract of hashset. I think there must be a way to use the hash of the value for the comparison.

**Comment by axal25** [comment]
Hi, I was wondering why all versions of javadocs state that contains method for HashSet needs only correct equal method to work. I'm probably doing something I should not do, but I have created a subclass of custom class and implemented equals method for both of them. Then I tried to use HashSet<ChildClass> and childClassSet.contains(parentClassObject) and it was returning false, even when they both were equal to each other (unit tested). I know it's common knowledge that both equals and hashCode methods should be implemented correctly for HashSet to work, but why is it stated in javadoc that "Returns true if this set contains the specified element. More formally, returns true if and only if this set contains an element e such that Objects.equals(o, e).". Shouldn't it be expanded on? Especially since it states "More formally" and it's actually more confusing since it come from official documentation.

PS: I'm sorry for posting offtopic, but I was hoping for clarification or correction if I'm misunderstanding something.

**Comment by joel-costigliola** [comment]
I don't know, I guess the javadoc authors were assuming that whenever `equals` is overridden `hashCode` should also be.

**Comment by idesofjackson** [comment]
Not sure if it is useful but per the javadocs:
Note: Great care must be exercised if mutable objects are used as set elements. The behavior of a set is not specified if the value of an object is changed in a manner that affects equals comparisons while the object is an element in the set. A special case of this prohibition is that it is not permissible for a set to contain itself as an element.

**Comment by mat3e** [comment]
Hi, today I also faced this issue.

I wanted to ensure `equals` and `hashCode` works correctly for my JPA entity. They were not, yet tests passed. // bigger issue

> [!NOTE]
> JPA entities have an extended lifecycle (e.g. can be loaded after app restarts), so it's tricky to implement `equals` and `hashCode` for them. Yet, e.g. for caching scenarios, `equals` and `hashCode` should be implemented. Relying on entity ID alone won't work if ID is generated and changes from `null` to the generated value after entity is persisted. And nice tests can help to guard this: https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/.

I was also showing how `equals` and `hashCode` can break reading from `HashSet` back when I was teaching Java. I didn't use AssertJ then, but it would be nice to show such things with fluent assertions. // small issue

I see some options for fixing were already listed in this thread
* `HashSet`-specific assertion // sounds nice IMO
* Dedicated method, like `containsWithHashCodeCheck(ELEMENT... values)`
* Warning when `contains()` and `actual.contains()` differ // IMO should happen if not other options

I also thought about `usingHashCode()` mode, similar to `usingRecursiveComparison()` and others 🤔

I started some hacking and  #3607 is my very draft PR. Initially I just wanted to override methods directly in my `HashSetAssert`, but then I checked `Iterables` and `Lists` and thought about adding `Sets`. Then, it appeared that `ComparisonStrategy` might be just enough. Let me know if and how to proceed 🤔

**Comment by joel-costigliola** [comment]
I think having specific assertions for HashSet is the way to go and the javadoc for contains should make it clear what the behavior is in this case.

**Comment by mat3e** [comment]
I think my current PR is fine. I wonder, but one can probably consider this as a breaking change, so `4.0.0-M1` might be good indeed 👍🏻

**Comment by scordio** [comment]
Hi @mat3e, I set `4.0.0-M1` mostly because `3.x` is officially in maintenance mode, and we don't plan to backport further changes unless they are major blockers.

**Comment by schiemon** [comment]
Hi all,
I stumbled across this issue when I noticed slow `assertThat(actualHashSet).contains(expectedValue)` calls on bigger `HashSet`s. Even more surprisingly I noticed that `containsAll` calls are slow too:

```java
package org.assertj.tests.core.perf;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

class ContainsWithAsPerfTest {
  @ParameterizedTest
  @CsvSource({
      "100000, 50000, true, true", // passes, <100ms
      "100000, 50000, true, false", // timeout
      "100000, 50000, false, true", // passes, <100ms
      "100000, 50000, false, false", // timeout
  })
  @Timeout(value = 10, unit = TimeUnit.SECONDS)
  void contains_on_large_hashset_with_as_description_should_be_fast(int actualSize, int expectedSize, boolean withContainsAll, boolean withNativeCheck) {
    HashSet<Integer> actualSet = new HashSet<>(actualSize);
    for (int i = 0; i < actualSize; i++) {
      actualSet.add(i);
    }

    Set<Integer> expectedSet = new HashSet<>();

    for (int i = 0; i < expectedSize; i++) {
      expectedSet.add(i);
    }

    if (withContainsAll)  {
      if  (withNativeCheck) {
        assertThat(actualSet.containsAll(expectedSet)).isTrue();
      } else {
        assertThat(actualSet).containsAll(expectedSet);
      }
    } else {
      for (Integer expectedValue: expectedSet) {

        if (withNativeCheck) {
          assertThat(actualSet.contains(expectedValue)).isTrue();
        } else {
          assertThat(actualSet).contains(expectedValue);
        }
      }
    }
  }
}
```

Here is a flamegraph from the last run (`HashSetAssert.contains`):

<img width="4902" height="1768" src="https://github.com/user-attachments/assets/b70716c1-8e16-4c6e-a053-893fe214591f" />

As you can see it comes from the linear `StandardComparisonStrategy.iterableContains`. Suppose I cannot reduce the `HashSet` what is the recommended API to use for this? Is it possible to skip the redundant `super.iterableContains` check?

**Comment by mat3e** [comment]
> Can we skip the hash check?

Yes, [skippingHashCodeComparison](https://github.com/assertj/assertj/blob/c9764fc664fc82899b5f6f761b3ede3e103fcc29/assertj-core/src/main/java/org/assertj/core/api/HashSetAssert.java#L66) is the way to go, e.g.

```java
then(actualSet).skippingHashCodeComparison().contains(expectedValue);
```

**Comment by scordio** [comment]
> Yes, `skippingHashCodeComparison` is the way to go

Such an API ~does not exist today...~ hasn't been released yet.

**Comment by scordio** [comment]
Thanks for the feedback, @schiemon!

I raised #4233 to analyze it further and continue the discussion.

**Comment by scordio** [comment]
Apologies @mat3e, I hadn't realized you were referring to what hasn't been released yet 🤦 

When adding `skippingHashCodeComparison` to the example above, all the tests were able to complete on my machine:

<img width="859" height="121" alt="Image" src="https://github.com/user-attachments/assets/4de809ab-8b2d-4ce2-941e-ce5e95ef0b09" />

@schiemon would you have a chance to [try out](https://central.sonatype.org/publish/publish-portal-snapshots/#consuming-snapshot-releases-for-your-project) `4.0.0-SNAPSHOT` and see if `skippingHashCodeComparison` satisfies your use case?

**Comment by mat3e** [comment]
> Apologies [@mat3e](https://github.com/mat3e), I hadn't realized you were referring to what hasn't been released yet 🤦

Yes, `HashSetAssert` as a feature targets version 4, but `skippingHashCodeComparison` is included since its beginning 👍🏻 

I'll take a quick look at #4233 though. Under the hood assertions iterate through all the elements, and we have quite some iterations in the provided example tests.

**Comment by schiemon** [comment]
> @schiemon would you have a chance to [try out](https://central.sonatype.org/publish/publish-portal-snapshots/#consuming-snapshot-releases-for-your-project) 4.0.0-SNAPSHOT and see if skippingHashCodeComparison satisfies your use case?

Just to answer this question: Yes,  I did a rerun with `skippingHashCodeComparison` and got similar duration times as you. As such, it is still very slow as with `skippingHashCodeComparison` we are still doing the linear scan as @mat3e mentioned. In terms of my use-case I am now just using the "native" `assertThat(actualSet.contains(expectedSet)).isTrue()` check instead which is unfortunate of course as I am not getting good debug information in case the assertion fails.
